DKMS build fails on Linux 6.10.3 with uart_state and circ_buf related errors

The module build fails on a recent kernel update. My previous kernel was 6.9.9, which works fine

Here's the build log:
```
DKMS make.log for xtrx-0.0.1+git20190320.5ae3a3e-3.5 for kernel 6.10.3-amd64 (x86_64)
Tue Aug 13 09:59:06 AM CEST 2024
make: Entering directory '/usr/src/linux-headers-6.10.3-amd64'
  CC [M]  /var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build/xtrx.o
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build/xtrx.c: In function ‘xtrx_uart_do_tx’:
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build/xtrx.c:472:28: error: ‘struct uart_state’ has no member named ‘xmit’
  472 |         xmit = &port->state->xmit;
      |                            ^~
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build/xtrx.c:473:13: error: implicit declaration of function ‘uart_circ_empty’; did you mean ‘uart_lsr_tx_empty’? [-Werror=implicit-function-declaration]
  473 |         if (uart_circ_empty(xmit))
      |             ^~~~~~~~~~~~~~~
      |             uart_lsr_tx_empty
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build/xtrx.c:480:25: error: invalid use of undefined type ‘struct circ_buf’
  480 |                 c = xmit->buf[xmit->tail] & 0xff;
      |                         ^~
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build/xtrx.c:480:35: error: invalid use of undefined type ‘struct circ_buf’
  480 |                 c = xmit->buf[xmit->tail] & 0xff;
      |                                   ^~
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build/xtrx.c:484:21: error: invalid use of undefined type ‘struct circ_buf’
  484 |                 xmit->tail = (xmit->tail + 1) & (UART_XMIT_SIZE - 1);
      |                     ^~
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build/xtrx.c:484:35: error: invalid use of undefined type ‘struct circ_buf’
  484 |                 xmit->tail = (xmit->tail + 1) & (UART_XMIT_SIZE - 1);
      |                                   ^~
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build/xtrx.c:490:13: error: implicit declaration of function ‘uart_circ_chars_pending’ [-Werror=implicit-function-declaration]
  490 |         if (uart_circ_chars_pending(xmit) < WAKEUP_CHARS)
      |             ^~~~~~~~~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make[2]: *** [/usr/src/linux-headers-6.10.3-common/scripts/Makefile.build:249: /var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build/xtrx.o] Error 1
make[1]: *** [/usr/src/linux-headers-6.10.3-common/Makefile:1959: /var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.5/build] Error 2
make: *** [/usr/src/linux-headers-6.10.3-common/Makefile:252: __sub-make] Error 2
make: Leaving directory '/usr/src/linux-headers-6.10.3-amd64'
```

--- a/xtrx.c
+++ b/xtrx.c
@@ -35,6 +35,7 @@
 #include <linux/serial.h>
 #include <linux/serial_core.h>
 #include <linux/serial_reg.h>
+#include <linux/kfifo.h>
 #include <linux/device.h>
 #include <linux/time.h>
 #include <linux/pps_kernel.h>
@@ -451,7 +452,11 @@ void xtrx_uart_do_rx(struct uart_port *port, unsigned* fifo_used)
 
 static void xtrx_uart_do_tx(struct uart_port *port, unsigned fifo_used)
 {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 10, 0)
 	struct circ_buf *xmit;
+#else
+	struct tty_port *tport;
+#endif
 	unsigned int max_count;
 	struct xtrx_dev *dev = xtrx_dev_from_uart_port(port);
 
@@ -469,29 +474,53 @@ static void xtrx_uart_do_tx(struct uart_port *port, unsigned fifo_used)
 		return;
 	}
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 10, 0)
 	xmit = &port->state->xmit;
 	if (uart_circ_empty(xmit))
 		goto txq_empty;
+#else
+	tport = &port->state->port;
+	if (kfifo_is_empty(&tport->xmit_fifo))
+		goto txq_empty;
+#endif
 
 	max_count = port->fifosize - fifo_used;
 	while (max_count--) {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 10, 0)
 		unsigned int c;
 
 		c = xmit->buf[xmit->tail] & 0xff;
 		//printk(KERN_NOTICE PFX "Char: %x\n", c);
+#else
+		unsigned char c;
+
+		if (kfifo_is_empty(&tport->xmit_fifo) || kfifo_get(&tport->xmit_fifo, &c) == 0)
+			break;
+#endif
 		xtrx_writel(dev, (port->line % XTRX_UART_NUM == XTRX_UART_LINE_GPS) ?
-				    GP_PORT_WR_UART_TX : GP_PORT_WR_SIM_TX, c);
+					GP_PORT_WR_UART_TX : GP_PORT_WR_SIM_TX, c);
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 10, 0)
 		xmit->tail = (xmit->tail + 1) & (UART_XMIT_SIZE - 1);
 		port->icount.tx++;
 		if (uart_circ_empty(xmit))
 			break;
+#endif
 	}
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 10, 0)
 	if (uart_circ_chars_pending(xmit) < WAKEUP_CHARS)
 		uart_write_wakeup(port);
 
 	if (uart_circ_empty(xmit))
 		goto txq_empty;
+#else
+	if (kfifo_len(&tport->xmit_fifo) < WAKEUP_CHARS)
+		uart_write_wakeup(port);
+
+	if (kfifo_is_empty(&tport->xmit_fifo))
+		goto txq_empty;
+#endif
 	return;
 
 txq_empty:

